src/nnp_model.F

```diff
@@ -0,0 +1,275 @@
+!--------------------------------------------------------------------------------------------------!
+
+! **************************************************************************************************
+!> \brief  Methods dealing with core routines for artificial neural networks
+!> \author Christoph Schran (christoph.schran@rub.de)
+!> \date   2020-10-10
+! **************************************************************************************************
+MODULE nnp_model
+
+   USE cp_log_handling,                 ONLY: cp_get_default_logger,&
+                                              cp_logger_get_default_unit_nr,&
+                                              cp_logger_type
+   USE cp_para_types,                   ONLY: cp_para_env_type
+   USE kinds,                           ONLY: default_string_length,&
+                                              dp,&
+                                              max_line_length
+   USE nnp_environment_types,           ONLY: &
+        nnp_actfnct_cos, nnp_actfnct_exp, nnp_actfnct_gaus, nnp_actfnct_invsig, nnp_actfnct_lin, &
+        nnp_actfnct_quad, nnp_actfnct_sig, nnp_actfnct_softplus, nnp_actfnct_tanh, nnp_arc_type, &
+        nnp_type
+#include "./base/base_uses.f90"
+
+   IMPLICIT NONE
+
+   PRIVATE
+
+   LOGICAL, PARAMETER, PRIVATE :: debug_this_module = .TRUE.
+   CHARACTER(len=*), PARAMETER, PRIVATE :: moduleN = 'nnp_model'
+
+   PUBLIC :: nnp_write_arc, &
+             nnp_predict, &
+             nnp_gradients
+
+CONTAINS
+
+! ***************************************************************************
+!> \brief Write neural network architecture information
+!> \param nnp ...
+!> \param para_env ...
+! **************************************************************************************************
+   SUBROUTINE nnp_write_arc(nnp, para_env)
+      TYPE(nnp_type), POINTER                            :: nnp
+      TYPE(cp_para_env_type), POINTER                    :: para_env
+
+      CHARACTER(len=*), PARAMETER :: routineN = 'nnp_write_arc', routineP = moduleN//':'//routineN
+
+      CHARACTER(len=default_string_length)               :: my_label
+      INTEGER                                            :: i, j, unit_nr
+      TYPE(cp_logger_type), POINTER                      :: logger
+
+      NULLIFY (logger)
+      logger => cp_get_default_logger()
+
+      my_label = "NNP| "
+      IF (para_env%ionode) THEN
+         unit_nr = cp_logger_get_default_unit_nr(logger)
+         DO i = 1, nnp%n_ele
+            WRITE (unit_nr, *) TRIM(my_label)//" Neural network specification for element "// &
+               nnp%ele(i)//":"
+            DO j = 1, nnp%n_layer
+               WRITE (unit_nr, '(1X,A,1X,I3,1X,A,1X,I2)') TRIM(my_label), &
+                  nnp%arc(i)%n_nodes(j), "nodes in layer", j
+            END DO
+         END DO
+      END IF
+
+      RETURN
+
+   END SUBROUTINE nnp_write_arc
+
+! ***************************************************************************
```

Comment:
Some of the Doxygen boxes have too short top lines. I guess, we have a but in our doxyfier script.